Add support for P4estMesh to `subcell-limiting`

bennibolm opened the pull request:
Adding support for subcell limiting with P4estMesh.

TODOs:

- [x] Check if StructuredMesh and P4estMesh give the same results
-> Different order in calculating of the boundary fluxes yields slightly different (machine precision) results

bennibolm committed:
Add support for P4estMesh to `subcell-limiting`

bennibolm commented on src/solvers/dgsem_p4est/dg_2d_subcell_limiters.jl:
```diff
@@ -0,0 +1,212 @@
+            flux_primary = flux(u_primary, normal_direction, equations)
+            flux_secondary = flux(u_secondary, normal_direction, equations)
+
+            # TODO: Correct order? flux_primary - flux_secondary
+            bar_state = 0.5 * (u_primary + u_secondary) -
+                        0.5 * (flux_primary - flux_secondary) / lambda
```
I'm not really sure about the order of these two fluxes here. @amrueda

amrueda replied:
This is a bit tricky because the bar state must be computed in the direction of the local coordinate system for the limiting procedure to make sense. Therefore, I believe that the calculation of the bar state has to be done twice. See https://github.com/bennibolm/Trixi.jl/pull/121#discussion_r1402078383.

bennibolm commented on src/solvers/dgsem_p4est/dg_2d_subcell_limiters.jl:
```diff
@@ -0,0 +1,212 @@
+                flux_inner = flux(u_inner, normal_direction, equations)
+                flux_outer = flux(u_outer, normal_direction, equations)
+
+                # TODO: Correct order?
+                bar_state = 0.5 * (u_inner + u_outer) -
+                            0.5 * (flux_inner - flux_outer) / lambda
```
Same as above

amrueda replied:
See https://github.com/bennibolm/Trixi.jl/pull/121#discussion_r1402078383.

bennibolm replied:
So, when your comment is correct, I have to just `flux_outer - flux_inner` here? Since the normal_vector is pointing outwards.

amrueda replied:
Ahh yes, there is only one outward-pointing normal direction at the boundaries. `flux_outer - flux_inner` should work.

bennibolm committed:
Add elixir with supersonic flow

bennibolm commented on examples/p4est_2d_dgsem/elixir_euler_supersonic_cylinder_sc_subcell.jl:
```diff
@@ -0,0 +1,138 @@
+#                            :Right => boundary_condition_inflow_outflow,
+#                            :Left => boundary_condition_inflow_outflow)
+
+boundary_conditions = Dict(:Bottom => boundary_condition_slip_wall,
+                           :Circle => boundary_condition_slip_wall,
+                           :Top => boundary_condition_slip_wall,
+                           :Right => boundary_condition_outflow,
+                           :Left => boundary_condition_supersonic_inflow)
```
One problem right now is that `boundary_condition_outflow` and `boundary_condition_supersonic_inflow` are defined in this elixir. Therefore, in the calculation of the bounds, the boundary outer state cannot be calculated with the [`get_boundary_outer_state`](https://github.com/bennibolm/Trixi.jl/blob/96ca8e25d5e1697107042b37a94c1dc525be5f38/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L1819-L1858) routine.

bennibolm replied:
So, I also tried it with the boundary condition using the characteristics, but this needed a few complicated changes for the new mesh type I have not implemented yet.

amrueda replied:
Can't this be fixed by defining the function `get_boundary_outer_state` for the new BCs? Can't this be done in the Elixir as well?

amrueda replied:
> I also tried it with the boundary condition using the characteristics.

The characteristics BC should work as both inflow and outflow.

bennibolm replied:
> Can't this be fixed by defining the function `get_boundary_outer_state` for the new BCs? Can't this be done in the Elixir as well?

That's a good idea. Thanks.

bennibolm replied:
Oh yes, that actually works. Thanks. Unfortunately, it doesn't change the failing simulation.

amrueda replied:
But the simulation fails only with the bar states, is that right?... Then the failure might be related to the computation of the bar states at the interfaces and boundaries.

bennibolm replied:
Yes, exactly.

bennibolm committed:
Extract inner code of `characteristic_boundary_value_function`

amrueda commented on src/solvers/dgsem_p4est/dg_2d_subcell_limiters.jl:
```diff
@@ -0,0 +1,212 @@
+            flux_secondary = flux(u_secondary, normal_direction, equations)
+
+            # TODO: Correct order? flux_primary - flux_secondary
+            bar_state = 0.5 * (u_primary + u_secondary) -
+                        0.5 * (flux_primary - flux_secondary) / lambda
+            if primary_direction == 1
+                set_node_vars!(bar_states1, bar_state, equations, dg, i_primary,
+                               j_primary, primary_element)
+            elseif primary_direction == 2
+                set_node_vars!(bar_states1, bar_state, equations, dg, i_primary + 1,
+                               j_primary, primary_element)
+            elseif primary_direction == 3
+                set_node_vars!(bar_states2, bar_state, equations, dg, i_primary,
+                               j_primary, primary_element)
+            else # primary_direction == 4
+                set_node_vars!(bar_states2, bar_state, equations, dg, i_primary,
+                               j_primary + 1, primary_element)
+            end
+            if secondary_direction == 1
+                set_node_vars!(bar_states1, bar_state, equations, dg, i_secondary,
+                               j_secondary, secondary_element)
+            elseif secondary_direction == 2
+                set_node_vars!(bar_states1, bar_state, equations, dg, i_secondary + 1,
+                               j_secondary, secondary_element)
+            elseif secondary_direction == 3
+                set_node_vars!(bar_states2, bar_state, equations, dg, i_secondary,
+                               j_secondary, secondary_element)
+            else # secondary_direction == 4
+                set_node_vars!(bar_states2, bar_state, equations, dg, i_secondary,
+                               j_secondary + 1, secondary_element)
+            end
```
The bar state must be computed with the "normal direction" of the local coordinate system. Therefore, in this case it is better to compute two bar states. One for the primary element and one for the secondary element. I.e., something along the lines of:
```suggestion
            bar_state_primary = 0.5 * (u_primary + u_secondary) -
                                0.5 * (flux_secondary - flux_primary) / lambda
            bar_state_secondary = 0.5 * (u_primary + u_secondary) -
                                  0.5 * (flux_primary - flux_secondary) / lambda
            if primary_direction == 1
                set_node_vars!(bar_states1, bar_state_primary, equations, dg, i_primary,
                               j_primary, primary_element)
            elseif primary_direction == 2
                set_node_vars!(bar_states1, bar_state_primary, equations, dg, i_primary + 1,
                               j_primary, primary_element)
            elseif primary_direction == 3
                set_node_vars!(bar_states2, bar_state_primary, equations, dg, i_primary,
                               j_primary, primary_element)
            else # primary_direction == 4
                set_node_vars!(bar_states2, bar_state_primary, equations, dg, i_primary,
                               j_primary + 1, primary_element)
            end
            if secondary_direction == 1
                set_node_vars!(bar_states1, bar_state_secondary, equations, dg, i_secondary,
                               j_secondary, secondary_element)
            elseif secondary_direction == 2
                set_node_vars!(bar_states1, bar_state_secondary, equations, dg, i_secondary + 1,
                               j_secondary, secondary_element)
            elseif secondary_direction == 3
                set_node_vars!(bar_states2, bar_state_secondary, equations, dg, i_secondary,
                               j_secondary, secondary_element)
            else # secondary_direction == 4
                set_node_vars!(bar_states2, bar_state_secondary, equations, dg, i_secondary,
                               j_secondary + 1, secondary_element)
            end
```
Can you please check if that works?

bennibolm replied:
We didn't do this for the implementation for the TreeMesh and StructuredMesh, did we? https://github.com/bennibolm/Trixi.jl/blob/ad64cd72897716a98b157d667a8a978bc137d958/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L701-L749

amrueda replied:
No, but in TreeMesh and StructuredMesh, the local coordinate systems are always aligned. I.e. xi+ of primary element points in the same direction as xi+ of the secondary element. The same for eta+. In general 
unstructured meshes, xi+ of the primary element can point in any direction of the secondary element (xi+, xi-, eta+, eta-).

amrueda replied:
I just realized that you might need something like
```julia
bar_state_secondary = 0.5 * (u_primary + u_secondary) -
                                  0.5 * secondary_direction * (flux_primary - flux_secondary) / lambda
```

bennibolm replied:
> No, but in TreeMesh and StructuredMesh, the local coordinate systems are always aligned. I.e. xi+ of primary element points in the same direction as xi+ of the secondary element. The same for eta+. In general
unstructured meshes, xi+ of the primary element can point in any direction of the secondary element (xi+, xi-, eta+, eta-).

I'm not completely sure what you mean and if your conclusion is right.
First, StructuredMesh is also curved. But okay, it's structured. So, the xi+ interface always correspond to the xi- interface of the neighbor element.

But I mean, also for our unstructured meshes, the normal vector of the primary element is the negative version of the normal vector of the secondary element at this node, right? So, I see the point of changing the order of the fluxes for both elements, but NOT only for P4estMesh but for StructuredMesh AS WELL.

bennibolm replied:
> `bar_state_secondary = 0.5 * (u_primary + u_secondary) -
                                  0.5 * secondary_direction * (flux_primary - flux_secondary) / lambda`

You mean to somehow add the normal vector here? Multiplying `primary_direction` doesn't really make sense, does it? It is just an Integer between 1 and 4.

amrueda replied:
Ah, sorry. I misinterpreted `direction` here. I mean something along the lines of:
```julia
if isodd(secondary_direction)
     bar_state_secondary = 0.5 * (u_primary + u_secondary) -
                                  0.5 * (flux_secondary - flux_primary) / lambda
else
    bar_state_secondary = 0.5 * (u_primary + u_secondary) -
                                  0.5 * (flux_primary - flux_secondary) / lambda
end
```

amrueda replied:
> But I mean, also for our unstructured meshes, the normal vector of the primary element is the negative version of the normal vector of the secondary element at this node, right?

Yes, that's correct. The difference is that this normal vector can be pointing in negative or positive xi or eta direction (with respect to the primary element's local coordinate system), and in negative or or positive xi or eta direction (with respect to the secondary element's local coordinate system).

> So, I see the point of changing the order of the fluxes for both elements, but NOT only for P4estMesh but for StructuredMesh AS WELL.

Since the local coordinate systems always match in StructuredMesh, a change is not necessary IMHO.

bennibolm replied:
> Ah, sorry. I misinterpreted `direction` here. I mean something along the lines of:
> 
> ```julia
> if isodd(secondary_direction)
>      bar_state_secondary = 0.5 * (u_primary + u_secondary) -
>                                   0.5 * (flux_secondary - flux_primary) / lambda
> else
>     bar_state_secondary = 0.5 * (u_primary + u_secondary) -
>                                   0.5 * (flux_primary - flux_secondary) / lambda
> end
> ```

Do I need this for `bar_state_primary` as well?

amrueda replied:
> > Ah, sorry. I misinterpreted `direction` here. I mean something along the lines of:
> > ```julia
> > if isodd(secondary_direction)
> >      bar_state_secondary = 0.5 * (u_primary + u_secondary) -
> >                                   0.5 * (flux_secondary - flux_primary) / lambda
> > else
> >     bar_state_secondary = 0.5 * (u_primary + u_secondary) -
> >                                   0.5 * (flux_primary - flux_secondary) / lambda
> > end
> > ```
> 
> Do I need this for `bar_state_primary` as well?

I believe something like that is already done inside get_normal_direction. We have to check if the sign is correct, but I assume it is.

bennibolm committed:
Add support for `BoundaryConditionDirichlet` for P4estMesh

bennibolm committed:
Fix calculation of `vn`

bennibolm committed:
Merge branch 'subcell-limiting' into subcell-limiting-p4est-all

bennibolm commented:
I'm thinking about whether it's useful to keep this default implementation of `get_boundary_outer_state` that only returns the inner value. When keeping it, the user doesn't really get an information if the used boundary condition has an influence to the bounds calculation or not.
https://github.com/bennibolm/Trixi.jl/blob/1b2ce66a1edfb816fcd43f27c9eada2b39d5f061/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L1853-L1858
What do you think about that @amrueda ?

EDIT: On the other hand, it is useful for boundary conditions like `boundary_condition_outflow`.

amrueda commented:
> I'm thinking about whether it's useful to keep this default implementation of `get_boundary_outer_state` that only returns the inner value. When keeping it, the user doesn't really get an information if the used boundary condition has an influence to the bounds calculation or not.
> 
> https://github.com/bennibolm/Trixi.jl/blob/1b2ce66a1edfb816fcd43f27c9eada2b39d5f061/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L1853-L1858
> 
> What do you think about that @amrueda ?

True, I was also thinking recently about this routine. I think it is better to remove it because it can lead a user to think that the bound computation takes into account their BCs. However, sometimes it makes sense to just use the inner solution. Is it possible to activate this computation with an input argument and keep it inactive as default?

bennibolm commented:
> True, I was also thinking recently about this routine. I think it is better to remove it because it can lead a user to think that the bound computation takes into account their BCs. However, sometimes it makes sense to just use the inner solution. Is it possible to activate this computation with an input argument and keep it inactive as default?

Yesterday, I revised the implementation of this routine. It is much more flexible and cleaner now.
Yes, it would be possible to add a boolean variable like `use_inner_value_as_boundary_value_as_default` (or something longer :wink: ) and throwing an error when the default routine is clled but the variable is `false`. But actually I think this is not very nice and we would have another stupid parameter :persevere:
So, I think I will just delete the default routine. If someone still wants to just use the inner value it is about 4 lines of code to add to the elixir.

bennibolm committed:
Remove default routine for `get_boundary_outer_state`

bennibolm committed:
Revise calculation of bar states

amrueda commented on src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl:
```diff
@@ -1839,21 +1842,15 @@ end
                                           direction, equations, dg, indices...)
-    u_rotate = rotate_to_x(u_inner, normal_direction, equations)
+    factor = (normal_direction[1] * u_inner[2] + normal_direction[2] * u_inner[3])
+    u_normal = factor * SVector(normal_direction / norm(normal_direction))
```
You need the square of the norm (`norm(normal_direction)^2 = sum(normal_direction^2)`) because you are multiplying twice by the normal vector. I believe the second variant is more efficient because `norm()` takes the square root.
```suggestion
    u_normal = factor * SVector(normal_direction / sum(normal_direction^2))
```

bennibolm replied:
Ah true, thanks.

bennibolm committed:
Fix allocations with `foreach(enumerate(...))`

bennibolm committed:
Fix normalization in `get_boundary_outer_state`

bennibolm committed:
Add TODO note for future

bennibolm committed:
Remove unnecessary code; format

bennibolm committed:
Adapt test errors after normalizing normal vectors

bennibolm committed:
Add dispatch for equations

bennibolm committed:
Merge branch 'subcell-limiting' into subcell-limiting-p4est-all

bennibolm committed:
Merge branch 'subcell-limiting' into subcell-limiting-p4est-all

bennibolm committed:
Fix normalization; Is done inside boundary_value_function

bennibolm committed:
Fix bug in calculation of bar state at interfaces for P4estMesh

bennibolm committed:
Add mesh as parameter of `get_boundary_outer_state`
- Fixes bug that P4estMesh calls the right `characteristic_boundary_state` routine

bennibolm committed:
Add p4est double mach elixir

bennibolm committed:
Fix error

bennibolm committed:
Merge branch 'main' into subcell-limiting-p4est-all

bennibolm committed:
Merge branch 'subcell-limiting' into subcell-limiting-p4est-all

bennibolm committed:
Merge branch 'subcell-limiting' into subcell-limiting-p4est-all

bennibolm committed:
Remove comment

bennibolm committed:
Merge branch 'subcell-limiting' into subcell-limiting-p4est-all

bennibolm committed:
Add test for supersonic flow

bennibolm committed:
fmt

bennibolm committed:
Add comments

amrueda commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -442,6 +444,24 @@ end
+                                                        equations)
+end
+
 # Inner function to distinguish between different mesh types.
 @inline function characteristic_boundary_value_function_inner(outer_boundary_value_function,
                                                               u_inner, srho, vn, x, t,
```
I find the name of this function a bit confusing. Someone could think that it returns the "inner" boundary value. I don't know what would be standard in Trixi , but I would rename it to something like:
```suggestion
# Function to compute the outer state of the characteristics-based boundary condition. This function is called by all mesh types.
@inline function calc_characteristic_boundary_value_function(outer_boundary_value_function,
                                                             u_inner, srho, vn, x, t,
```

bennibolm replied:
Good point. Thanks

amrueda reviewed (commented):
LGTM!
Just a minor comment:

bennibolm committed:
Update src/equations/compressible_euler_2d.jl

Co-authored-by: Andrés Rueda-Ramírez <aruedara@uni-koeln.de>

bennibolm committed:
fmt

bennibolm committed:
Fix error
